### ISSUE TITLE
Question regarding Algorithm 1: Compliance loss computation 

### ISSUE THREAD

**Issue by gd193** [UW-ERSL/TOuNN#4]
Hello, 

I've been reading your TOuNN paper and played a bit with the code and some parameters and got some cool results.

I do however have a question regarding the loss computation. In Algorithm 1, Line 8, you state that you compute the compliance loss as 

$L_c = \frac{\sum_{e} \rho_{e}^{p} J_{e}}{J_0}$

In the code, you however compute the compliance loss as 

`torch.sum(torch.div(Jelem,nn_rho**self.FE.penal))/self.obj0; # compliance` 

so the formula would be $L_c = \frac{1}{J_0} \sum_{e} \frac{  J_{e}}{\rho_{e}^{p}}$.

How does this equal the compliance, which is stated in (4) to be $J = \sum_e \rho_{e}^p J_e$

Is `Jelem`not $J_e$ ? Or am I missing something else?

Kind regards

**Comment by aadityacs** [comment]
This was a workaround we implemented back when we did not have a solver that was also written in pytorch ( and thus needed a workaround to compute the gradients).

I would prolly check out the JAX code https://github.com/UW-ERSL/JAXTOuNN where we don't these workarounds .

Cheers

**Comment by gd193** [comment]
Workaround what exactly?

Your FE solver returns the $J_e$, then you sum with the $\rho_e$ to get the compliance and backprop from that. I just stuggle to understand why `torch.div` 

I don't know JAX that well, that's why I was reading your older pytorch code

**Comment by aadityacs** [comment]
Well Je is a function of rho. And for optimization we need dJ_drho. However with the FE solver not written using pytorch, we wouldn't be able to compute the sensitivity using autograd. And hence the workaround. The right way to approach this would be to also write the solver using torch. We didn't have that when we wrote this paper. Subsequently, we wrote a differentiable solver in JAX. 

For a much cleaner and updated pytorch code you could see
https://colab.research.google.com/drive/1Mtgxga1vyZDYHud0Djut1AJLBSew_NOI?usp=sharing
